File: src/ios/AES256.swift
import Foundation
import CommonCrypto

// La classe AES256, che estende CDVPlugin, espone le funzioni encrypt e decrypt
@objc(AES256)
class AES256: CDVPlugin {
    
    // MARK: - Parametri di inizializzazione PBKDF2 e AES
    // I parametri di derivazione della chiave devono essere identici in entrambi gli ambienti.
    
    /// Numero di iterazioni per la derivazione (PBKDF2)
    private static let iterations: Int = 10000
    /// Salt utilizzato per la derivazione; deve essere lo stesso in iOS ed Android per l’interoperabilità
    private static let salt: String = "someSaltValue"
    /// Algoritmo usato per PBKDF2 (HMAC-SHA256)
    private static let algorithm: String = "sha256" // Informativo: usato in CommonCrypto
    /// Lunghezza della chiave derivata in byte (256 bit = 32 byte)
    private static let keyLength: Int = 32
    /// Lunghezza dell’IV (AES block size: 16 byte)
    private static let ivLength: Int = 16

    /// Coda di esecuzione asincrona per le operazioni crittografiche
    private static let aes256Queue = DispatchQueue(label: "AESQUEUE", qos: .background, attributes: .concurrent)
    
    
    // MARK: - Funzioni di interfaccia Cordova

    /// Metodo di encrypt chiamato da JavaScript
    @objc(encrypt:)
    func encrypt(command: CDVInvokedUrlCommand) {
        AES256.aes256Queue.async {
            guard let args = command.arguments as? [Any],
                  args.count >= 2,
                  let plaintext = args[0] as? String,
                  let password = args[1] as? String else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Parametri non validi")
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                return
            }
            if let encrypted = self.encryptText(plaintext: plaintext, password: password) {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: encrypted)
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            } else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Cifratura fallita")
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
        }
    }
    
    /// Metodo di decrypt chiamato da JavaScript
    @objc(decrypt:)
    func decrypt(command: CDVInvokedUrlCommand) {
        AES256.aes256Queue.async {
            guard let args = command.arguments as? [Any],
                  args.count >= 2,
                  let ciphertext = args[0] as? String,
                  let password = args[1] as? String else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Parametri non validi")
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                return
            }
            if let decrypted = self.decryptText(ciphertext: ciphertext, password: password) {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: decrypted)
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            } else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Decifratura fallita")
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
        }
    }
    
    
    // MARK: - Funzioni interne per encrypt e decrypt

    /// Cifra il testo in chiaro usando AES256 in modalità CBC con padding PKCS7.
    /// - Parametri:
    ///   - plaintext: Il testo in chiaro da cifrare.
    ///   - password: La password da cui derivare chiave ed IV.
    /// - Restituisce: Il testo cifrato in Base64 oppure nil in caso di errore.
    func encryptText(plaintext: String, password: String) -> String? {
        // Deriva chiave e IV usando PBKDF2
        guard let keyAndIV = self.deriveKeyAndIV(password: password) else { return nil }
        let key = keyAndIV.key
        let iv = keyAndIV.iv
        
        guard let dataToEncrypt = plaintext.data(using: .utf8) else { return nil }
        let bufferSize = dataToEncrypt.count + kCCBlockSizeAES128
        var buffer = Data(count: bufferSize)
        var numBytesEncrypted: size_t = 0
        
        // Il padding PKCS7 è usato e garantisce interoperabilità
        let cryptStatus = key.withUnsafeBytes { keyBytes in
            iv.withUnsafeBytes { ivBytes in
                dataToEncrypt.withUnsafeBytes { dataBytes in
                    buffer.withUnsafeMutableBytes { bufferBytes in
                        CCCrypt(CCOperation(kCCEncrypt),
                                CCAlgorithm(kCCAlgorithmAES),
                                CCOptions(kCCOptionPKCS7Padding),
                                keyBytes.baseAddress, key.count,
                                ivBytes.baseAddress,
                                dataBytes.baseAddress, dataToEncrypt.count,
                                bufferBytes.baseAddress, bufferSize,
                                &numBytesEncrypted)
                    }
                }
            }
        }
        
        if cryptStatus == kCCSuccess {
            buffer.count = numBytesEncrypted
            return buffer.base64EncodedString()
        } else {
            return nil
        }
    }
    
    /// Decifra il testo cifrato in Base64 usando AES256 in modalità CBC con padding PKCS7.
    /// - Parametri:
    ///   - ciphertext: Il testo cifrato in Base64.
    ///   - password: La password da cui derivare chiave ed IV.
    /// - Restituisce: Il testo decifrato oppure nil in caso di errore.
    func decryptText(ciphertext: String, password: String) -> String? {
        guard let keyAndIV = self.deriveKeyAndIV(password: password) else { return nil }
        let key = keyAndIV.key
        let iv = keyAndIV.iv
        
        guard let dataToDecrypt = Data(base64Encoded: ciphertext) else { return nil }
        let bufferSize = dataToDecrypt.count + kCCBlockSizeAES128
        var buffer = Data(count: bufferSize)
        var numBytesDecrypted: size_t = 0
        
        let cryptStatus = key.withUnsafeBytes { keyBytes in
            iv.withUnsafeBytes { ivBytes in
                dataToDecrypt.withUnsafeBytes { dataBytes in
                    buffer.withUnsafeMutableBytes { bufferBytes in
                        CCCrypt(CCOperation(kCCDecrypt),
                                CCAlgorithm(kCCAlgorithmAES),
                                CCOptions(kCCOptionPKCS7Padding),
                                keyBytes.baseAddress, key.count,
                                ivBytes.baseAddress,
                                dataBytes.baseAddress, dataToDecrypt.count,
                                bufferBytes.baseAddress, bufferSize,
                                &numBytesDecrypted)
                    }
                }
            }
        }
        
        if cryptStatus == kCCSuccess {
            buffer.count = numBytesDecrypted
            return String(data: buffer, encoding: .utf8)
        } else {
            return nil
        }
    }
    
    /// Deriva una chiave e un IV utilizzando PBKDF2 con HMAC-SHA256.
    /// - Parametro password: La password di input.
    /// - Restituisce: Una tupla contenente (key, iv) oppure nil in caso di errore.
    func deriveKeyAndIV(password: String) -> (key: Data, iv: Data)? {
        // I parametri di derivazione (iterations, salt, keyLength, ivLength) sono DEVE ESSERE IDENTICI in entrambi gli ambienti!
        guard let passwordData = password.data(using: .utf8),
              let saltData = AES256.salt.data(using: .utf8) else { return nil }
        
        var derivedBytes = Data(count: AES256.keyLength + AES256.ivLength)
        let derivationStatus = derivedBytes.withUnsafeMutableBytes { derivedBytesPtr in
            passwordData.withUnsafeBytes { passwordBytes in
                saltData.withUnsafeBytes { saltBytes in
                    CCKeyDerivationPBKDF(CCPBKDFAlgorithm(kCCPBKDF2),
                                         password, passwordData.count,
                                         saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self), saltData.count,
                                         CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                                         UInt32(AES256.iterations),
                                         derivedBytesPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                         derivedBytes.count)
                }
            }
        }
        if derivationStatus != kCCSuccess {
            return nil
        }
        let key = derivedBytes.subdata(in: 0..<AES256.keyLength)
        let iv = derivedBytes.subdata(in: AES256.keyLength..<(AES256.keyLength + AES256.ivLength))
        return (key, iv)
    }
}

File: PHP/AES256.php
<?php

class AES256 {
    // Parametri per PBKDF2 e AES
    private static $PBKDF2_ITERATIONS = 10000;
    private static $PBKDF2_SALT = "someSaltValue";
    private static $PBKDF2_ALGORITHM = "sha256"; // In PHP, per PBKDF2 si usa "sha256" per HMAC-SHA256
    private static $AES_KEY_LENGTH = 256; // in bit (256 bit = 32 byte)
    private static $IV_LENGTH = 16;       // 16 byte per AES (block size)

    /**
     * Cifra il testo in chiaro usando AES-256-CBC.
     *
     * @param string $password La password da cui derivare la chiave.
     * @param string $plaintext Il testo in chiaro da cifrare.
     * @param string $iv Il vettore di inizializzazione (IV) (deve essere lungo 16 byte).
     * @return string Il testo cifrato codificato in Base64.
     * @throws Exception
     */
    public static function encrypt($password, $plaintext, $iv) {
         // Deriva la chiave (e l'IV interno, non usato in seguito) dalla password
         $keyAndIV = self::deriveKeyAndIV($password);
         $key = $keyAndIV['key'];

         // Si utilizza il IV fornito come parametro, che deve essere una stringa di 16 byte.
         $encrypted = openssl_encrypt($plaintext, 'AES-256-CBC', $key, OPENSSL_RAW_DATA, $iv);
         if ($encrypted === false) {
              throw new Exception("Encryption failed");
         }
         return base64_encode($encrypted);
    }

    /**
     * Decifra il testo cifrato usando AES-256-CBC.
     *
     * @param string $password La password da cui derivare la chiave.
     * @param string $ciphertext Il testo cifrato codificato in Base64.
     * @param string $iv Il vettore di inizializzazione (IV) (deve essere lungo 16 byte).
     * @return string Il testo in chiaro decifrato.
     * @throws Exception
     */
    public static function decrypt($password, $ciphertext, $iv) {
         $keyAndIV = self::deriveKeyAndIV($password);
         $key = $keyAndIV['key'];

         $decodedCiphertext = base64_decode($ciphertext);
         $decrypted = openssl_decrypt($decodedCiphertext, 'AES-256-CBC', $key, OPENSSL_RAW_DATA, $iv);
         if ($decrypted === false) {
              //throw new Exception("Decryption failed");
              $decrypted='n.a.';
         }
         return $decrypted;
    }

    /**
     * Genera la "secure key" derivata dalla password usando PBKDF2.
     *
     * @param string $password La password.
     * @return string La secure key in formato esadecimale.
     * @throws Exception
     */
    public static function generateSecureKey($password) {
         $keyAndIV = self::deriveKeyAndIV($password);
         return bin2hex($keyAndIV['key']);
    }

    /**
     * Genera la "secure IV" derivata dalla password usando PBKDF2.
     *
     * @param string $password La password.
     * @return string La secure IV in formato esadecimale.
     * @throws Exception
     *
     * @note Nel codice Java originale anche questo metodo restituisce il valore derivato dalla chiave.
     * Se si desidera restituire l'IV derivato, modificare il return in:
     * return bin2hex($keyAndIV['iv']);
     */
    public static function generateSecureIV($password) {
         $keyAndIV = self::deriveKeyAndIV($password);
         return bin2hex($keyAndIV['key']);
    }

    /**
     * Deriva la chiave AES e l'IV dalla password usando PBKDF2 con HMAC-SHA256.
     *
     * @param string $password La password.
     * @return array Un array associativo contenente 'key' e 'iv'.
     * @throws Exception
     */
    private static function deriveKeyAndIV($password) {
         // Lunghezza totale in byte = chiave (32 byte) + IV (16 byte) = 48 byte.
         $totalLength = (self::$AES_KEY_LENGTH / 8) + self::$IV_LENGTH; // 32 + 16 = 48

         // Deriva il materiale della chiave con PBKDF2
         $derived = hash_pbkdf2(
             self::$PBKDF2_ALGORITHM,
             $password,
             self::$PBKDF2_SALT,
             self::$PBKDF2_ITERATIONS,
             $totalLength,
             true  // output in formato binario (raw)
         );

         if ($derived === false || strlen($derived) !== $totalLength) {
              throw new Exception("Key derivation failed");
         }
         $key = substr($derived, 0, self::$AES_KEY_LENGTH / 8);
         $iv = substr($derived, self::$AES_KEY_LENGTH / 8, self::$IV_LENGTH);
         return ['key' => $key, 'iv' => $iv];
    }
}

/*
----------------------------
   Esempio di utilizzo:

try {
    $password  = "password123";
    $plaintext = "Akuna matata: tutta frenesia";
    $iv        = "1234567890123456"; // Deve essere lungo 16 byte


    echo "Testo da cifrare: ".$plaintext."\n";
    echo "Password: ".$password."\n";
    echo "IV: ".$iv."\n\n";

    // Cifratura
    $ciphertext = AES256::encrypt($password, $plaintext, $iv);
    echo "Testo cifrato: " . $ciphertext . "\n";

    // Decifratura
    $decrypted = AES256::decrypt($password, $ciphertext, $iv);
    echo "Testo decifrato: " . $decrypted . "\n";

    // Genera secure key e secure IV
    echo "Secure Key: " . AES256::generateSecureKey($password) . "\n";
    echo "Secure IV: "  . AES256::generateSecureIV($password) . "\n";
} catch (Exception $e) {
    echo "Errore: " . $e->getMessage() . "\n";
}
-------------------------------
*/


?>
